track_analyzing/log_parser.cpp

```diff
@@ -0,0 +1,168 @@
+#include "track_analyzing/log_parser.hpp"
+
+#include "generator/borders_generator.hpp"
+#include "generator/borders_loader.hpp"
+
+#include "platform/platform.hpp"
+
+#include "coding/file_name_utils.hpp"
+#include "coding/hex.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "base/timer.hpp"
+
+#include <cstdint>
+#include <fstream>
+#include <regex>
+#include <unordered_set>
+
+using namespace std;
+using namespace track_analyzing;
+
+namespace
+{
+vector<DataPoint> ReadDataPoints(string const & data)
+{
+  string const decoded = FromHex(data);
+  vector<DataPoint> points;
+  MemReader memReader(decoded.data(), decoded.size());
+  ReaderSource<MemReader> src(memReader);
+  coding::TrafficGPSEncoder::DeserializeDataPoints(1 /* version */, src, points);
+  return points;
+}
+
+class PointToMwmId final
+{
+public:
+  PointToMwmId(shared_ptr<m4::Tree<routing::NumMwmId>> mwmTree,
+               routing::NumMwmIds const & numMwmIds, string const & dataDir)
+    : m_mwmTree(mwmTree)
+  {
+    numMwmIds.ForEachId([&](routing::NumMwmId numMwmId) {
+      string const & mwmName = numMwmIds.GetFile(numMwmId).GetName();
+      string const polyFile = my::JoinPath(dataDir, BORDERS_DIR, mwmName + BORDERS_EXTENSION);
+      osm::LoadBorders(polyFile, m_borders[numMwmId]);
+    });
+  }
+
+  routing::NumMwmId FindMwmId(m2::PointD const & point, routing::NumMwmId expectedId) const
+  {
+    if (expectedId != routing::kFakeNumMwmId && m2::RegionsContain(GetBorders(expectedId), point))
+      return expectedId;
+
+    routing::NumMwmId result = routing::kFakeNumMwmId;
+    m2::RectD const rect = MercatorBounds::RectByCenterXYAndSizeInMeters(point, 1);
+    m_mwmTree->ForEachInRect(rect, [&](routing::NumMwmId numMwmId) {
+      if (m2::RegionsContain(GetBorders(numMwmId), point))
+      {
+        result = numMwmId;
+        return;
```

Comment:
If you wish to do early exit, please, modify ForEachInRect to support lambdas that returns bools.